### ISSUE TITLE
build failure due to invalid merkle tree entries with dynamic execution

### ISSUE THREAD

**Issue by kevingessner** [bazelbuild/bazel#12401]
### Description of the problem / feature request:

We use remote execution (engflow) for our builds. I'm experimenting with dynamic execution, but getting an intermittent crash. From my investigation, this appears to be caused by Bazel reading invalid file contents for action inputs merkle tree entries. I've not seen this with non-dynamic remote execution. I'm using `--internal_spawn_scheduler --local_cpu_resources=HOST_CPUS*0.75 --experimental_local_execution_delay=250` -- I am using the default legacy spawn scheduler because my build fails completely with `--nolegacy_spawn_scheduler`.

On failure, Engflow returns an error like

```
INVALID_ARGUMENT: Invalid argument: resource_name=25461dbfb431cb48e30118a0277671de8e8a56a3b1198492e24b2fb605d7575e/351575 - received digest does not match computed digest 5a79696cab7720f5d399895b5e8fc170fca60ef8bd6fb5b5f7f3bbb247e8954b/351575
```

This means that when Engflow is validating an uploaded artifact, it doesn't match the declared hash. It doesn't fail on every build, but more often on ones with lots of new artifacts (i.e. low cache hit rate). It's always a different hash for both the expected and actual. Objects with the expected hash do exist and are valid in the remote cache; the invalid hash is not a valid object (i.e. it's not a different object).

I wanted to verify that this wasn't corruption introduced by the network, so I hacked in some code to make Bazel validate merkle tree entries and uploads ([patch](https://gist.github.com/kevingessner/cb93fcdaee9961183fd2f8fb9ef27a8c)). This shows that the corruption is happening inside of Bazel: it is creating invalid digests that don't match the file contents. Here's an example of such a failure:

```
merkle path [redacted]/com/github/scopt/scopt_2.12/4.0.0-RC2/scopt_2.12-4.0.0-RC2.jar mismatch d19a4e8b8c013a56e03bc57bdf87abe6297c974cf907585d00284eae61c6ac91 != ab8e7ebffbf2807fd7173db8d0151bafe63f23952094366dd8a31f111955c027```
```

The digest log shows that the file was added by `DirectoryTreeBuilder`:
```
adding input bazel-out/host/bin/redacted/com/github/scopt/scopt_2.12/4.0.0-RC2/scopt_2.12-4.0.0-RC2.jar = File:[[<execution_root>]bazel-out/host/bin]redacted/com/github/scopt/scopt_2.12/4.0.0-RC2/scopt_2.12-4.0.0-RC2.jar = RegularFileArtifactValue{digest=ab8e7ebffbf2807fd7173db8d0151bafe63f23952094366dd8a31f111955c027, size=32768, proxy=ctime of 1603487999260 and nodeId of 260049459}
/scopt_2.12-4.0.0-RC2.jar = File:[[<execution_root>]bazel-out/host/bin]redacted/com/github/scopt/scopt_2.12/4.0.0-RC2/scopt_2.12-4.0.0-RC2.jar = RegularFileArtifactValue{digest=ab8e7ebffbf2807fd7173db8d0151bafe63f23952094366dd8a31f111955c027, size=32768, proxy=ctime of 1603487999260 and nodeId of 260049459}
```

It reports finding the file being exactly 32768 bytes, but the file is 105131 bytes (The file is https://repo1.maven.org/maven2/com/github/scopt/scopt_2.12/4.0.0-RC2/scopt_2.12-4.0.0-RC2.jar if you want to try this at home). And the digest is wrong; `d19a4e8b8c013a56e03bc57bdf87abe6297c974cf907585d00284eae61c6ac91` is the correct sha256sum of the full JAR.

So Bazel is reading a truncated version of the file! The hash of the first 32KiB of the file is the incorrect one Bazel reports:

```
head -c 32768 bazel-out/host/bin/redacted/com/github/scopt/scopt_2.12/4.0.0-RC2/scopt_2.12-4.0.0-RC2.jar | sha256sum
ab8e7ebffbf2807fd7173db8d0151bafe63f23952094366dd8a31f111955c027  -
```

So, in summary: With dynamic execution, Bazel intermittently reads an invalid size and digest for action inputs. This causes the build to fail if the whole file is later uploaded to the remote executor, as the invalid digest does not match.

I'm not familiar enough with action input and tree construction to dig deeper into why this bad read is happening.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Unfortunately, this bug is intermittent and nondeterministic. I'm trying to work up a self-contained example but not having any luck.
 
### What operating system are you running Bazel on?

Centos 7 with a 5.1 kernel

### What's the output of `bazel info release`?

`development version`

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

`bazel build //src:bazel-dev` on top of the 3.5.0 tag with these patches: https://gist.github.com/kevingessner/cb93fcdaee9961183fd2f8fb9ef27a8c

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

```
:) > git remote get-url origin ; git rev-parse master ; git rev-parse HEAD
https://github.com/bazelbuild/bazel
d22bb66eb60a7799c9222798862940b2fa8dc33b
76cc17fc449099d7e3f14b6e5d0ac4f9ee6bb7cf
```

###  Have you found anything relevant by searching the web?

No

**Comment by larsrc-google** [comment]
@ulfjack Ulf, you know more about the Merkle trees than I do. 

Kevin: I would be very interested in hearing how the new scheduler fails.

**Comment by ulfjack** [comment]
Thanks, @larsrc-google. I've seen the issue, but haven't had time yet to investigate. It's very, very odd.

@kevingessner, did you see https://github.com/bazelbuild/bazel/issues/12364? Also, can you say if it's always 32kbyte, or is it sometimes more or less than that?

**Comment by kevingessner** [comment]
@ulfjack No, it's not always 32KiB. In fact, I just had a failure where the file was added with the correct size but wrong hash:

```
:( > ack scala-library-2.12.12.jar digests.log
adding input bazel-out/host/bin/external/.../org/scala-lang/scala-library/2.12.12/scala-library-2.12.12.jar = File:[[<execution_root>]bazel-out/host/bin]external/.../org/scala-lang/scala-library/2.12.12/scala-library-2.12.12.jar = RegularFileArtifactValue{digest=864c7c01ced4c425219ed9a4d53dda4adb92fb56e9a5bfe86e924b166c580711, size=5432322, proxy=ctime of 1604507699834 and nodeId of 1199659375}
adding input bazel-out/host/bin/external/.../org/scala-lang/scala-library/2.12.12/scala-library-2.12.12.jar = File:[[<execution_root>]bazel-out/host/bin]external/.../org/scala-lang/scala-library/2.12.12/scala-library-2.12.12.jar = RegularFileArtifactValue{digest=864c7c01ced4c425219ed9a4d53dda4adb92fb56e9a5bfe86e924b166c580711, size=5432322, proxy=ctime of 1604507699834 and nodeId of 1199659375}

:) > ll bazel-out/host/bin/external/.../org/scala-lang/scala-library/2.12.12/scala-library-2.12.12.jar
-r-xr-xr-x. 1 kgessner etsians 5432322 Nov  4 16:35 bazel-out/host/bin/external/.../org/scala-lang/scala-library/2.12.12/scala-library-2.12.12.jar

:) > sha256sum bazel-out/host/bin/external/.../org/scala-lang/scala-library/2.12.12/scala-library-2.12.12.jar
1673ffe8792021f704caddfe92067ed1ec75229907f84380ad68fe621358c925  bazel-out/host/bin/external/.../org/scala-lang/scala-library/2.12.12/scala-library-2.12.12.jar
```

**Comment by kevingessner** [comment]
@larsrc-google With `--nolegacy_spawn_scheduler`, I get the same failure as #12364: `FileNotFoundException` with the DynamicSpawnStrategy/RemoteCache stack trace, e.g.

```
Building external/io_bazel_rules_scala/src/java/io/bazel/rulesscala/worker/libworker.jar (1 source file) failed (Exit 34): java.io.FileNotFoundException: /home/.../worker/libworker-native-header.jar.tmp (No such file or directory)                                                                                                                                   at com.google.devtools.build.lib.unix.NativePosixFiles.lstat(Native Method)                                                              at com.google.devtools.build.lib.unix.UnixFileSystem.statInternal(UnixFileSystem.java:186)                                               at com.google.devtools.build.lib.unix.UnixFileSystem.stat(UnixFileSystem.java:176)                                                       at com.google.devtools.build.lib.vfs.Path.stat(Path.java:418)                                                                            at com.google.devtools.build.lib.vfs.FileSystemUtils.moveFile(FileSystemUtils.java:454)                                                  at com.google.devtools.build.lib.remote.RemoteCache.moveOutputsToFinalLocation(RemoteCache.java:415)                                     at com.google.devtools.build.lib.remote.RemoteCache.download(RemoteCache.java:374)                                                       at com.google.devtools.build.lib.remote.RemoteSpawnRunner.downloadAndFinalizeSpawnResult(RemoteSpawnRunner.java:440)                     at com.google.devtools.build.lib.remote.RemoteSpawnRunner.exec(RemoteSpawnRunner.java:266)                                               at com.google.devtools.build.lib.exec.SpawnRunner.execAsync(SpawnRunner.java:240)                                                        at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:132)                                         at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy.runRemotely(DynamicSpawnStrategy.java:397)                                 at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy.access$400(DynamicSpawnStrategy.java:64)                                   at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy$2.callImpl(DynamicSpawnStrategy.java:298)                                  at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy$Branch.call(DynamicSpawnStrategy.java:471)                                 at com.google.devtools.build.lib.dynamic.DynamicSpawnStrategy$Branch.call(DynamicSpawnStrategy.java:408)                                 at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)                                                                                                                            at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)                                                    at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by larsrc-google** [comment]
@kevingessner #12364 is now closed, with that fix do you still get the above error if you run with the new scheduler?

**Comment by kevingessner** [comment]
@larsrc-google Yes, that patch does appear to fix dynamic exec with the new scheduler! I cherry-picked it onto 3.5.0 and my build worked. Do you expect that fix to be in 3.8?

Trying to repro this merkle tree problem with the new scheduler if I can.

**Comment by larsrc-google** [comment]
Yes, that fix should be in the upcoming 4.0.

**Comment by larsrc-google** [comment]
Actually, it's already in 3.7.
